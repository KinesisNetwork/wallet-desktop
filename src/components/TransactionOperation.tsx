import {
  CreateAccountOperationRecord,
  OperationRecord,
  PaymentOperationRecord,
} from 'js-kinesis-sdk'
import * as React from 'react'

import { AddressView } from '@components/AddressView'
import { HorizontalLabelledField } from '@components/LabelledField'
import { AddressDisplay, Currency, TransactionOperationView } from '@types'

export interface Props {
  transactionWithOperation: TransactionOperationView
  currency: Currency
}

const isTransfer = (
  op: OperationRecord,
): op is PaymentOperationRecord | CreateAccountOperationRecord =>
  ['create_account', 'payment'].includes(op.type)

const getAddress = (t: TransactionOperationView) => {
  switch (t.operation.type) {
    case 'create_account':
      return t.isIncoming ? t.operation.funder : t.operation.account
    case 'payment':
      return t.isIncoming ? t.operation.from : t.operation.to
    default:
      return t.source
  }
}

const getAmount = (t: TransactionOperationView) => {
  let amount = 0
  switch (t.operation.type) {
    case 'create_account':
      amount = Number(t.operation.starting_balance)
      break
    case 'payment':
      amount = Number(t.operation.amount)
      break
    default:
      return amount
  }
  return t.isIncoming ? amount.toFixed(5) : (amount + Number(t.fee)).toFixed(5)
}

const isCurrencyOperation = (amount: number | string, currency: Currency) => {
  return amount && `${amount} ${currency}`
}

const TransactionIcon: React.SFC<{ t: TransactionOperationView }> = ({ t }) =>
  isTransfer(t.operation) ? (
    t.isIncoming ? (
      <span className="icon is-large has-text-success">
        <i className="fal fa-lg fa-arrow-down" />
      </span>
    ) : (
      <span className="icon is-large has-text-danger">
        <i className="fal fa-lg fa-arrow-up" />
      </span>
    )
  ) : (
    <span className="icon is-large has-text-grey-light">
      <i className="fal fa-lg fa-dash" />
    </span>
  )

interface StateProps {
  moreInfoIsHidden: boolean
  toggleMoreInfo: () => any
}
const TransactionCard: React.SFC<Props & StateProps> = ({
  transactionWithOperation: t,
  currency,
  moreInfoIsHidden,
  toggleMoreInfo,
}) => (
  <article className="level">
    <div
      className="columns box level-item is-gapless has-text-grey-lighter is-multiline"
      style={{ width: '100%' }}
    >
      <div className="column is-4">
        <div className="is-flex" style={{ alignItems: 'center' }}>
          <TransactionIcon t={t} />
          <span className="has-text-weight-bold">
            <AddressView address={getAddress(t)} addressDisplay={AddressDisplay.payee} />
          </span>
        </div>
      </div>
      <div className="column is-3">
        <p>{t.memo || '(no description)'}</p>
      </div>
      <div className="column is-2">
        <span className="icon is-small has-text-success">
          <i className="fal fa-xs fa-check-circle" />
        </span>
        <span>{t.date.toTimeString().slice(0, 8)}</span>
      </div>
      <div className="column is-2 has-text-weight-bold has-text-right">
        <span className={`has-text-${t.isIncoming ? 'success' : 'danger'}`}>
          {isCurrencyOperation(getAmount(t), currency)}
        </span>
      </div>
      <div className="column is-1 has-text-right">
        <button className="button is-text" onClick={toggleMoreInfo}>
          {moreInfoIsHidden && (
            <span className="icon">
              <i className="fal fa-lg fa-angle-down" />
            </span>
          )}
          {!moreInfoIsHidden && (
            <span className="icon">
              <i className="fal fa-lg fa-angle-up" />
            </span>
          )}
        </button>
      </div>
      <div className={`column is-12 ${moreInfoIsHidden ? 'is-hidden' : ''}`}>
        <HorizontalLabelledField
          label={t.isIncoming ? "Sender's Address:" : "Recipient's Address:"}
          value={getAddress(t)}
          isCompact={true}
        />
        <HorizontalLabelledField
          label="Transaction Hash:"
          value={t.operation.transaction_hash}
          isCompact={true}
        />
        <HorizontalLabelledField label="Fee:" value={`${t.fee} ${currency}`} isCompact={true} />
      </div>
    </div>
  </article>
)

interface State {
  moreInfoIsHidden: boolean
}
class TransactionCardStateful extends React.Component<Props, State> {
  state: State = {
    moreInfoIsHidden: true,
  }

  render() {
    return (
      <TransactionCard
        moreInfoIsHidden={this.state.moreInfoIsHidden}
        toggleMoreInfo={this.toggleMoreInfo}
        {...this.props}
      />
    )
  }

  private toggleMoreInfo = () =>
    this.setState(state => ({ moreInfoIsHidden: !state.moreInfoIsHidden }))
}

export { TransactionCardStateful as TransactionCard }
