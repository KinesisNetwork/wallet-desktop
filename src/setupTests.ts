import { configure } from 'enzyme'
import * as Adapter from 'enzyme-adapter-react-16'

export default configure({ adapter: new Adapter() })
