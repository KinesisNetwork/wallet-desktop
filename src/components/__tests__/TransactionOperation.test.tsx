import { shallow } from 'enzyme';
import { startCase } from 'lodash'
import * as React from 'react'

import { TransactionCard } from '@components/TransactionOperation'
import '../../setupTests'

describe('TransactionCard', () => {
  let props

  beforeEach(() => {
    props = {
      transactionWithOperation: {
        date: '',
        fee: '0.00046',
        isIncoming: true,
        operation: {
          account: 'qwer',
          created_at: '2018-08-23T03:41:54Z',
          funder: 'asdf',
          id: '1234',
          source_account: 'zxcv',
          starting_balance: '0.1',
          transaction_hash: '785d',
          succeeds: () => null,
          type: 'create_account'
        }
      }
    }
  })

  it('renders correctly', () => {
    const wrapper = shallow(<TransactionCard {...props} />)

    expect(wrapper).toMatchSnapshot()
  })

  it('displays an up arrow icon and a header matching operation type', () => {
    const wrapper = shallow(<TransactionCard {...props} />)

    const messageHeader = wrapper.find('.message-header')

    expect(messageHeader).toHaveLength(1)
    expect(messageHeader.find('i').hasClass('fa-arrow-up')).toBe(true)
    expect(messageHeader.text()).toBe('Create Account')
  })

  it('renders eight HorizontalLabelledField components as renderOperationRecords filters operations', () => {
    const wrapper = shallow(<TransactionCard {...props} />)

    const horizontalLabelledField = wrapper.find('HorizontalLabelledField')

    expect(horizontalLabelledField).toHaveLength(8)

    const labels = horizontalLabelledField.map((node) => node.prop('label'))

    expect(labels.includes(startCase('succeeds'))).toBe(false)
    expect(labels.includes(startCase('id'))).toBe(false)
    expect(labels.includes(startCase('type'))).toBe(false)
    expect(labels.includes(startCase('funder'))).toBe(true)
  })
})
