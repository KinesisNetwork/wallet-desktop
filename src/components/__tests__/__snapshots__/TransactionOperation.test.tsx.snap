// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionCard renders correctly 1`] = `
<article
  className="message is-small is-transparent"
>
  <div
    className="message-header"
  >
    <p>
      <span
        className="icon has-text-danger"
      >
        <i
          className="fas fa-arrow-up"
        />
      </span>
      <span>
        Create Account
      </span>
    </p>
  </div>
  <div
    className="message-body"
  >
    <HorizontalLabelledField
      isCompact={true}
      key="account"
      label="Account"
      value="qwer"
    />
    <HorizontalLabelledField
      isCompact={true}
      key="created_at"
      label="Created At"
      value="2018-08-23T03:41:54Z"
    />
    <HorizontalLabelledField
      isCompact={true}
      key="funder"
      label="Funder"
      value="asdf"
    />
    <HorizontalLabelledField
      isCompact={true}
      key="source_account"
      label="Source Account"
      value="zxcv"
    />
    <HorizontalLabelledField
      isCompact={true}
      key="starting_balance"
      label="Starting Balance"
      value="0.1"
    />
    <HorizontalLabelledField
      isCompact={true}
      key="transaction_hash"
      label="Transaction Hash"
      value="785d"
    />
    <HorizontalLabelledField
      isCompact={true}
      label="Fee"
      value="0.00046"
    />
    <HorizontalLabelledField
      isCompact={true}
      label="Memo"
    />
  </div>
</article>
`;
